docs/rfcs/034-storage-controller-restarts.md

```diff
@@ -0,0 +1,208 @@
+The aspiring leader loads the observed state into memory and the start-up sequence proceeds as usual,
+but *without* querying the pageservers in order to build up the observed state.
+
+The final start-up step is to update the `leader` database table as described in the [Database Table For Leader Synchronization](database-table-for-leader-synchronization)
+section. If this step fails, the storage controller process exits.
+
+Note that no row will exist in the `leaders` table for the first graceful restart. In that case, force update the `leader` table
+(without the WHERE clause) and perform with the pre-existing start-up procedure (i.e. build observed state by querying pageservers).
+
+[1] The API call might fail because there's no storage controller running (i.e. [restart](#storage-controller-crash-or-restart)),
+so we don't want to extend the unavailability period by much. We still want to retry since that's not the common case.
+
+### Handling Failures
+
+#### Storage Controller Crash Or Restart
+
+The storage controller may crash or be restarted outside of roll-outs. When a new pod is created, its call to
+`/step_down` will fail since the previous leader is no longer reachable. In this case perform the pre-existing
+start-up procedure and update the leader table (with the WHERE clause). If the update fails, the storage controller
+exists and consistency is maintained.
+
+#### Previous Leader Crashes Before New Leader Readiness
+
+When the previous leader (P1) crashes before the new leader (P2) passses the readiness check, Kubernetes will
+reconcile the old replica set and create a new pod for it (P1'). The `/step_down` API call will fail for P1'
+(see [2]).
+
+Now we have two cases to consider:
+* P2 updates the `leader` table first: The database update from P1' will fail and P1' will exit, or be terminated
+by Kubernetes depending on timings.
+* P1' updates the `leader` table first: The `hostname` field of the `leader` row stays the same, but the `start_timestamp` field changes.
+The database update from P2 will fail (since `start_timestamp` does not match). P2 will exit and Kubernetes will
+create a new replacement pod for it (P2'). Now the entire dance starts again, but with P1' as the leader and P2' as the incumbent.
+
+[2] P1 and P1' may (more likely than not) be the same pod and have the same hostname. The implementation
+should avoid this self reference and fail the API call at the client if the persisted hostname matches
+the current one.
+
+#### Previous Leader Crashes After New Leader Readiness
+
+The deployment's replica sets already satisfy the deployment's replica count requirements and the
+Kubernetes deployment rollout will just clean up the dead pod.
+
+#### New Leader Crashes Before Pasing Readiness Check
+
+The deployment controller scales up the new replica sets by creating a new pod. The entire procedure is repeated
+with the new pod.
+
+#### Network Partition Between New Pod and Previous Leader
```

Comment:
I would be a lot more keen to implement this if we were not running in k8s.

Even with a time based lease there may be scenarios in which there's two controllers acting as leaders for a _bounded_ amount of time (e.g. network partition). We get the same behaviour from k8s killing pods to bring the number of running replicas to 1 (the amount of time is still probably _bounded_, but it's trickier to figure out what that is).

So I'm not sure what this extra complexity gets us.

Reply:
In the case of a network partition, AIUI the kubelet doesn't kill pods -- a network partitioned kubernetes node can keep running its pods forever.  So I don't think we would have a bound on the time that two leaders were operating in this case?

A lease in our database could be checked by the storage controller periodically, and it can kill itself if/when it either can't reach the database, or finds that some other node has become leader.

Reply:
(this shouldn't be a sticking point though: it would be fine to implement what's described in this RFC, as if we do decide we want time leases we can extend it with that later)

Reply:
> In the case of a network partition, AIUI the kubelet doesn't kill pods -- a network partitioned kubernetes node can keep running its pods forever.

Ah I see what you mean now. I was thinking of network partition between the k8s pods, but yes if the controller is somehow partitioned there's no bound indeed. I'll think about it. Leaning toward including it in the RFC and implementing later down the line.

Reply:
I added a section describing the various safety measures we've discussed in [91b4533](https://github.com/neondatabase/neon/pull/8310/commits/91b4533358dfb00e06a15f4894de91c637d07065). I don't expect we'll implement all of them, but rather choose what feels more appropriate once the implementation is complete.